hal/armv7m/imxrt/117x/imxrt.c

```diff
@@ -389,6 +415,481 @@ __attribute__((section(".noxip"))) int _imxrt_setLevelLPCG(int clock, int level)
+}
+
+
+static u32 _imxrt_deinitArmPll(void)
+{
+	u32 reg = *(imxrt_common.anadig_pll + arm_pll_ctrl) & ~(1uL << 29u);
+
+	/* Disable and gate clock if not already */
+	if ((reg & ((1uL << 13u) | (1uL << 14u))) != 0u) {
+		/* Power down the PLL, disable clock */
+		reg &= ~((1uL << 13u) | (1uL << 14u));
+		/* Gate the clock */
+		reg |= 1uL << 30u;
+		*(imxrt_common.anadig_pll + arm_pll_ctrl) = reg;
+
+		hal_cpuDataSyncBarrier();
+		hal_cpuInstrBarrier();
+	}
+
+	return reg;
+}
+
+
+static int _imxrt_initArmPll(u8 loopDivider, u8 postDivider)
+{
+	u32 reg;
+
+	/*
+	 * Fin = XTALOSC = 24MHz
+	 * Fout = Fin * (loopDivider / (2 * postDivider))
+	 */
+
+	if ((loopDivider < 104u) || (208u < loopDivider)) {
+		return -1;
+	}
+
+	reg = _imxrt_deinitArmPll();
+
+	/* Set the configuration. */
+	reg &= ~((3uL << 15u) | 0xffu);
+	reg |= ((u32)(loopDivider & 0xffu) | (((u32)postDivider & 3uL) << 15u)) | (1uL << 30u) | (1uL << 13u);
+	*(imxrt_common.anadig_pll + arm_pll_ctrl) = reg;
+
+	hal_cpuDataSyncBarrier();
+	hal_cpuInstrBarrier();
+
+	_imxrt_delay(300u * 1000u);
+
+	/* Wait for stable PLL */
+	while ((*(imxrt_common.anadig_pll + arm_pll_ctrl) & (1uL << 29u)) == 0u) {
+	}
+
+	/* Enable the clock. */
+	reg |= 1uL << 14u;
+
+	/* Ungate the clock */
+	reg &= ~(1uL << 30u);
+
+	*(imxrt_common.anadig_pll + arm_pll_ctrl) = reg;
+
+	return 0;
+}
+
+
+static void _imxrt_initSysPll3(void)
+{
+	u32 reg;
+
+	/* check if configuration is the same, then only enable clock */
+	if ((*(imxrt_common.anadig_pll + sys_pll3_ctrl) & (1uL << 21u)) != 0u) {
+		/* if clock disable -> enable it */
+		if ((*(imxrt_common.anadig_pll + sys_pll3_ctrl) & (1uL << 13u)) == 0u) {
+			*(imxrt_common.anadig_pll + sys_pll3_ctrl) |= (1uL << 13u);
+		}
+
+		/* if clock is gated -> ungate */
+		if ((*(imxrt_common.anadig_pll + sys_pll3_ctrl) & (1uL << 30u)) != 0u) {
+			*(imxrt_common.anadig_pll + sys_pll3_ctrl) &= ~(1uL << 30u);
+		}
+
+		return;
+	}
+
+	/* Gate all PFDs */
+	*(imxrt_common.anadig_pll + sys_pll3_pfd) |= (1uL << 31u) | (1uL << 23u) | (1uL << 15u) | (1uL << 7u);
+
+	/* Enable, but gate clock */
+	reg = (1u << 4u) | (1uL << 30u);
+	*(imxrt_common.anadig_pll + sys_pll3_pfd) = reg;
+
+	_imxrt_delay(300u * 1000u);
+
+	/* Power off and hold ring off */
+	reg |= (1uL << 21u) | (1uL << 11u);
+	*(imxrt_common.anadig_pll + sys_pll3_pfd) = reg;
+
+	_imxrt_delay(300u * 1000u);
+
+	/* Deassert hold ring off */
+	reg &= ~(1uL << 11u);
+	*(imxrt_common.anadig_pll + sys_pll3_pfd) = reg;
+
+	/* Wait for stable PLL */
+	while ((*(imxrt_common.anadig_pll + sys_pll3_ctrl) & (1uL << 29u)) == 0u) {
+	}
+
+	/* Enable system pll3 and div2 clocks */
+	reg |= (1uL << 13u) | (1uL << 3u);
+
+	/* un-gate sys pll3 */
+	reg &= ~(1uL << 30u);
+	*(imxrt_common.anadig_pll + sys_pll3_pfd) = reg;
+}
+
+
+int _imxrt_setPfdPllFracClock(u8 pfd, u8 clk_pll, u8 frac)
+{
+	volatile u32 *ctrl;
+	volatile u32 *update;
+	volatile u32 stable;
+	u32 fracval;
+	u8 gatedpfd;
+
+	if ((frac < 12u) || (35u < frac)) {
```

Comment:
It seems that 12 is supported only for sys_pll3_pfd3.

Reply:
Right, thanks for spotting. Done.